nixos/modules/services/databases/postgresql.nix

```diff
@@ -36,6 +43,104 @@ let
 
   groupAccessAvailable = versionAtLeast postgresql.version "11.0";
 
+  upgradeScript =
+    pkgs.writeShellApplication {
+      name = "upgrade.sh";
+      text = with cfg.upgradeFrom; ''
+        # There is no old version to upgrade from - noop
+        if ! [[ -d "${settings.old-datadir}" ]]; then
+          echo "No old datadir to upgrade from (expecting ${settings.old-datadir})"
+          exit 0
+        fi
+
+        if [[ -f "${settings.old-datadir}/.upgraded" ]]; then
+          echo "Old data dir found, but was already migrated. Please remove the old path '${settings.old-datadir}' to suppress this message." >&2
+          exit 0
+        fi
+
+        pushd "${settings.old-datadir}"
+        ${settings.new-bindir}/pg_upgrade ${lib.cli.toGNUCommandLineShell {} settings}
+        popd
+        touch "${settings.new-datadir}/.post_upgrade" "${settings.old-datadir}/.upgraded"
+      '';
+    };
+
+  upgradeOptions = upgrade: {
+    options = {
+      package = mkOption {
+        type = types.package;
+        default = pkgs."postgresql_${defaultPackageVersion}";
+        defaultText = literalExpression ''pkgs.postgresql_${defaultPackageVersion}'';
+        description = lib.mdDoc "Package to upgrade from";
+      };
+
+      settings = mkOption {
+        type = types.submodule {
+          freeformType = types.attrsOf (types.oneOf [
+            types.bool
+            types.str
+            types.path
+            (types.listOf lib.types.str)
+          ]) // {
+            description = lib.mdDoc
+              "Type of flags accepted by pg_upgrade (see [pg_upgrade](https://www.postgresql.org/docs/current/pgupgrade.html))";
+          };
+
+          options = {
+            old-datadir = mkOption {
+              type = types.path;
+              default = "/var/lib/postgresql/${upgrade.config.package.psqlSchema}";
+              defaultText = literalExpression
+                ''"/var/lib/postgresql/''${config.services.postgresql.upgradeFrom.package.psqlSchema}"'';
+              example = "/mnt/postgresql/13";
+              description = lib.mdDoc ''
+                Location of the existing data directory.
+                It won't be deleted in the upgrade process.
+                See the [pg_upgrade docs](https://www.postgresql.org/docs/current/pgupgrade.html).
+              '';
+            };
+
+            new-datadir = mkOption {
+              type = types.path;
+              default = cfg.dataDir;
+              defaultText = literalExpression "config.services.postgresql.dataDir";
+              example = "/mnt/postgresql/15";
+              description = lib.mdDoc ''
+                The desired destination for the upgraded data directory.
+                See the [pg_upgrade docs](https://www.postgresql.org/docs/current/pgupgrade.html).
+              '';
+            };
+
+            old-bindir = mkOption {
+              type = types.path;
```

Comment:
We really need a type for that, @ncfavier do you have opinions?

Reply:
I've run out of opinions about Nix paths but @roberth might still have some.

I'll just mention that this feels like a "one more lane will fix it" situation: we already have `str`, `path`, `pathInStore` and `package` and it's already not always clear when to use which.

Reply:
<something something nix path processing something>

Reply:
> I'll just mention that this feels like a "one more lane will fix it" situation

Still curious in what @roberth has to say about this, but fwiw for me the most important thing is to not use `types.path` here because that can become a source of hard to grasp errors easily.

Reply:
Never write an absolute path literal.

Reply:
Agreed!

But since it'd be possible and is an easy mistake, we should still prohibit it on a typesystem-level here %) 

If I had to write a path type for this case here, I'd only allow strings (prefixed with `/` or something similar).

Reply:
> I'll just mention that this feels like a "one more lane will fix it" situation

I don't think we're inducing demand, and if we do, it's a good thing.
A type for non-read-only system paths would make a lot of sense to me.

In this specific instance though, we shouldn't ask for a bindir, but just a package, or even better, don't ask anything at all because it's better to track this in the system state where the data lives.